archive/imo/imo2008_q2.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2021 Manuel Candales. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Manuel Candales
+-/
+import data.real.basic
+import data.set.basic
+import data.set.finite
+
+/-!
+# IMO 2008 Q2
+(a) Prove that
+          ```x^2 / (x-1)^2 + y^2 / (y-1)^2 + z^2 / (z-1)^2 ≥ 1```
+for all real numbers x, y, z, each different from 1, and satisfying xyz = 1.
+
+(b) Prove that equality holds above for infinitely many triples of rational numbers x, y, z, each
+different from 1, and satisfying xyz = 1.
+
+# Solution
+(a) Since xyz = 1, we can apply the substitution x = a/b, y = b/c, z = c/a.
+Then we define m = c-b, n = b-a and rewrite the inequality as LHS - 1 ≥ 0
+using c, m and n. We factor LHS - 1 as a square, which finishes the proof.
+
+(b) We present a set W of rational triples. We prove that W is a subset of the
+set of rational solutions to the equation, and that W is infinite.
+-/
+
+lemma subst_abc {x y z : ℝ} (h : x*y*z = 1) :
+  ∃ a b c : ℝ, a ≠ 0 ∧ b ≠ 0 ∧ c ≠ 0 ∧ x = a/b ∧ y = b/c ∧ z = c /a :=
+begin
+  use [x, 1, 1/y],
+  have h₁ : x ≠ 0,              exact left_ne_zero_of_mul (left_ne_zero_of_mul_eq_one h),
+  have h₂ : (1 : ℝ) ≠ (0 : ℝ),  exact one_ne_zero,
+  have hy_ne_zero : y ≠ 0,      exact right_ne_zero_of_mul (left_ne_zero_of_mul_eq_one h),
+  have h₃ : 1/y ≠ 0,            exact one_div_ne_zero hy_ne_zero,
+  have h₄ : x = x / 1,          exact (div_one x).symm,
+  have h₅ : y = 1 / (1 / y),    exact (one_div_one_div y).symm,
+  have h₆ : z = 1 / y / x,      { field_simp, linarith [h] },
+  exact ⟨h₁, h₂, h₃, h₄, h₅, h₆⟩,
+end
+
+theorem imo2008_q2a (x y z : ℝ) (h : x*y*z = 1) (hx : x ≠ 1) (hy : y ≠ 1) (hz : z ≠ 1) :
+  x^2 / (x-1)^2 + y^2 / (y-1)^2 + z^2 / (z-1)^2 ≥ 1 :=
+begin
+  obtain ⟨a, b, c, ha, hb, hc, hx₂, hy₂, hz₂⟩ := subst_abc h,
+
+  set m := c-b with hm_abc,
+  set n := b-a with hn_abc,
+  have ha_cmn : a = c - m - n,      linarith,
```

Comment:
Here and below, when new goals are introduced, we usually isolate them with these "focusing" braces:
```suggestion
  have ha_cmn : a = c - m - n,      { linarith },
```

Reply:
done!